omnigibson/systems/system_base.py

```diff
@@ -134,6 +432,569 @@ def __init__(self):
+
+    @classmethod
+    def remove_particle_by_name(cls, name):
+        """
+        Remove particle with name @name from both the simulator and internal state
+
+        Args:
+            name (str): Name of the particle to remove
+        """
+        raise NotImplementedError()
+
+    @classmethod
+    def remove_all_group_particles(cls, group):
+        """
+        Remove particle with name @name from both the simulator as well as internally
+
+        Args:
+            group (str): Name of the attachment group to remove all particles from
+        """
+        # Make sure the group exists
+        cls._validate_group(group=group)
+        # Remove all particles from the group
+        for particle_name in tuple(cls._group_particles[group].keys()):
+            cls.remove_particle_by_name(name=particle_name)
+
+    @classmethod
+    def num_group_particles(cls, group):
+        """
+        Gets the number of particles for the given group in the simulator
+
+        Args:
+            group (str): Name of the attachment group to remove all particles from.
+
+        Returns:
+            int: Number of particles allocated to this group in the scene. Note that if @group does not
+                exist, this will return 0
+        """
+        # Make sure the group exists
+        cls._validate_group(group=group)
+        return len(cls._group_particles[group])
+
+    @classmethod
+    def get_group_name(cls, obj):
+        """
+        Grabs the corresponding group name for object @obj
+
+        Args:
+            obj (BaseObject): Object for which its procedurally generated particle attachment name should be grabbed
+
+        Returns:
+            str: Name of the attachment group to use when executing commands from this class on
+                that specific attachment group
+        """
+        return obj.name
+
+    @classmethod
+    def create_attachment_group(cls, obj):
+        """
+        Creates an attachment group internally for object @obj. Note that this does NOT automatically generate particles
+        for this object (should call generate_group_particles(...) ).
+
+        Args:
+            obj (BaseObject): Object for which a new particle attachment group will be created for
+
+        Returns:
+            str: Name of the attachment group to use when executing commands from this class on
+                that specific attachment group
+        """
+        group = cls.get_group_name(obj=obj)
+        # This should only happen once for a single attachment group, so we explicitly check to make sure the object
+        # doesn't already exist
+        assert group not in cls.groups, \
+            f"Cannot create new attachment group because group with name {group} already exists!"
+
+        # Create the group
+        cls._group_particles[group] = dict()
+        cls._group_objects[group] = obj
+
+        # Compute the group scale
+        cls._group_scales[group] = cls._compute_relative_group_scales(group=group)
```

Comment:
Should we only do this if cls.scale_relative_to_parent is True?

Reply:
done